### PR TITLE
Implement module mapper file for MSVC and Clang

### PR DESCRIPTION
**Pull request by Arthapz**
https://github.com/xmake-io/xmake/pull/2641

### CODE REVIEW

**Comment by Arthapz** [comment]
MSVC is done, i'm working on clang

**Comment by Arthapz** [comment]
Ok i think it's done

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 1]
```diff
@@ -27,31 +27,71 @@ import("utils.progress")
...
+    if not os.isfile(mapper_file) then
+        io.writefile(mapper_file, "")
+    end
+    return mapper_file
+end
+
+-- add a module into the mapper
+--
+-- e.g
+-- -fmodule-file=foo=build/.gens/Foo/rules/modules/cache/foo.ifc
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:42]: .ifc?

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 2]
```diff
@@ -399,9 +449,24 @@ function get_scandependenciesflag(target)
...
         end
         _g.scandependenciesflag = scandependenciesflag or false
     end
     return scandependenciesflag or nil
 end
+
+-- append .obj to final link
+function append_headerunits_objectfiles(target)
+    local linker_file = _get_linker_argument_file()
+
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:462]: please remove empty line

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 3]
```diff
@@ -399,9 +449,24 @@ function get_scandependenciesflag(target)
...
+    for line in io.lines(linker_file) do
+        if target:kind() == "binary" then
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:464]: target:is_binary()

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 4]
```diff
@@ -399,9 +449,24 @@ function get_scandependenciesflag(target)
...
+            target:add("ldflags", line, {force = true})
+        elseif target:kind() == "static" then
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:466]: target:is_static()

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 5]
```diff
@@ -399,9 +449,24 @@ function get_scandependenciesflag(target)
...
+            target:add("arflags", line, {force = true})
+        elseif target:kind() == "shared" then
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:468]: target:is_shared

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 6]
```diff
@@ -27,31 +27,71 @@ import("utils.progress")
...
-    target:add("cxxflags", prebuiltmodulepathflag .. cachedir, prebuiltmodulepathflag .. stlcachedir, {force = true})
-
-    -- add module cachedirs of all dependent targets with modules
-    -- this target maybe does not contain module files, @see https://github.com/xmake-io/xmake/issues/1858
-    for _, dep in ipairs(target:orderdeps()) do
-        cachedir = common.modules_cachedir(dep)
-        target:add("cxxflags", prebuiltmodulepathflag .. cachedir, {force = true})
-    end
+    target:add("cxxflags", noimplicitmodulemapsflag, {force = true})
+    -- target:add("cxxflags", prebuiltmodulepathflag .. cachedir, prebuiltmodulepathflag .. stlcachedir, {force = true})
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:92]: please remove unused line

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 7]
```diff
@@ -293,6 +330,34 @@ function get_modulesflag(target)
...
+        _g.builtinmodulemapflag = builtinmodulemapflag or false
+    end
+    return builtinmodulemapflag or nil
+end
+
+function get_modulemapfileflag(target)
+    local modulemapfileflag = _g.modulemapfileflag
+    if modulemapfileflag == nil then
+        local compinst = target:compiler("cxx")
+        local mapper_file = path.join(os.tmpdir(), "mapper.txt")
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:350]:
```
local mapper_file = path.join(os.tmpfile(), "mapper.txt")
```

os.tmpdir() is a fixed path and there are conflicts between different projects.

Comment (waruqi) [reply xmake/rules/c++/modules/modules_support/clang.lua:350]: why not use `build` directory?

File: xmake/rules/c++/modules/modules_support/gcc.lua [hunk 8]
```diff
@@ -165,14 +165,15 @@ function generate_stl_headerunits_for_batchcmds(target, batchcmds, headerunits,
...
+    for _, headerunit in ipairs(headerunits) do
         local bmifile = path.join(stlcachedir, headerunit.name .. get_bmi_extension())
-        if _add_module_to_mapper(mapper_file, headerunit.path, path.absolute(bmifile, projectdir)) then
+        if not os.isfile(bmifile) then
             local args = { "-c", "-x", "c++-system-header", headerunit.name }
-
             batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.headerunit.bmi %s", headerunit.name)
             batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), args))
         end
+
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/gcc.lua:175]: batchcmds:add_depfiles is missing

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 9]
```diff
@@ -399,9 +449,23 @@ function get_scandependenciesflag(target)
...
+            scandependenciesflag = "-scanDependencies"
         end
         _g.scandependenciesflag = scandependenciesflag or false
     end
     return scandependenciesflag or nil
 end
+
+-- append .obj to final link
+function append_headerunits_objectfiles(target)
+    local linker_file = _get_linker_argument_file()
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:461]: Why use file? Can't we just cache them in memory using target:data_set() or memcache()?

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/msvc.lua:461]: other targets may need them

Comment (waruqi) [reply xmake/rules/c++/modules/modules_support/msvc.lua:461]: > other targets may need them so for now, i add all the header unit objectfiles to all targets

you can use `common.memcache()`

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/msvc.lua:462]: memcache is clear between two invocation of xmake b ?

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/msvc.lua:462]: i need to use globalcache i think

Comment (waruqi) [reply xmake/rules/c++/modules/modules_support/msvc.lua:462]: why? Will all projects share it?

Comment (waruqi) [reply xmake/rules/c++/modules/modules_support/msvc.lua:462]: >     local linker_file = path.join(config.buildir(), "linkargs.txt")

I see that you are only using the local cache

File: xmake/rules/c++/modules/modules_support/msvc.lua [hunk 10]
```diff
@@ -131,32 +190,31 @@ function generate_stl_headerunits_for_batchcmds(target, batchcmds, headerunits,
...
+        local objectfile = bmifile .. ".obj"
+        if not os.isfile(bmifile) or not os.isfile(objectfile) then
+            local args = {headernameflag .. ":angle", headerunit.name, ifcoutputflag, stlcachedir, "-Fo" .. objectfile}
             batchcmds:show_progress(opt.progress, "${color.build.object}generating.cxx.headerunit.bmi %s", headerunit.name)
-            batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), args), {envs = vcvars})
-        end
-
-        local flag = {headerunitflag .. ":angle", headerunit.name .. "=" .. headerunit.name .. get_bmi_extension()}
-        table.join2(flags, flag)
+            batchcmds:vrunv(compinst:program(), table.join(compinst:compflags({target = target}), common_args, args), {envs = vcvars})
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/msvc.lua:201]: batchcmds:add_depfiles?

File: xmake/rules/c++/modules/modules_support/msvc.lua (hunk 9 above)
Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/msvc.lua:462]: why localcache doesn't store anything ?

```lua
    local cache = common.localcache():get("module_mapper") or {}
    for _, flag in ipairs(cache) do
        if flag:startswith(argument .. " " .. module) then
            return
        end
    end
    table.insert(cache, format("%s %s", argument, module))
    common.localcache():set("module_mapper", cache)
    common.localcache():save("module_mapper")
```

Comment (waruqi) [reply xmake/rules/c++/modules/modules_support/msvc.lua:464]: .xmake/macosx/x86_64/cache/cxxmodules

Comment (Arthapz) [reply xmake/rules/c++/modules/modules_support/msvc.lua:464]: nvm i was using memcache instead of the localcache in build_modules =P

**Comment by waruqi** [comment]
clang.lua is broken.

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 11]
```diff
@@ -27,31 +27,71 @@ import("utils.progress")
...
+    local mapflag = format("%s%s", modulefileflag, bmi)
+    if table.contains(cache, mapflag) then
+        return
+    end
+    table.insert(cache, mapflag)
+    common.localcache():set("mapflags", cache)
+    common.localcache():save("mapflags")
+end
+
+    local modulefileflag = get_modulefileflag(target)
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:64]: broken function

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 12]
```diff
@@ -225,14 +264,10 @@ function build_modules_for_batchcmds(target, batchcmds, objectfiles, modules, op
...
-    local modulefileflag = get_modulefileflag(target)
 
-    -- append deps modules
-    for _, dep in ipairs(target:orderdeps()) do
-        local flags = dep:data("cxx.modules.flags")
-        if flags then
-            target:add("cxxflags", flags, {force = true, expand = false})
-        end
+    -- append module mapper flags
+    for line in io.lines(mapper_file) do
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:269]: undef mapper_file

File: xmake/rules/c++/modules/modules_support/clang.lua [hunk 13]
```diff
@@ -293,6 +327,34 @@ function get_modulesflag(target)
...
+        _g.builtinmodulemapflag = builtinmodulemapflag or false
+    end
+    return builtinmodulemapflag or nil
+end
+
+function get_modulemapfileflag(target)
+    local modulemapfileflag = _g.modulemapfileflag
+    if modulemapfileflag == nil then
+        local compinst = target:compiler("cxx")
+        local mapper_file = path.join(os.tmpfile(), "mapper.txt")
```
Comment (waruqi) [inline xmake/rules/c++/modules/modules_support/clang.lua:347]: use mapper_file or localcache?

**Pull request merged**